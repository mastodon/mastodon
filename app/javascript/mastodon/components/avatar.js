import React from 'react';
import PropTypes from 'prop-types';
import ImmutablePropTypes from 'react-immutable-proptypes';
import { autoPlayGif } from '../initial_state';
import classNames from 'classnames';

export default class Avatar extends React.PureComponent {

  static propTypes = {
    account: ImmutablePropTypes.map,
    size: PropTypes.number.isRequired,
    style: PropTypes.object,
    inline: PropTypes.bool,
    animate: PropTypes.bool,
  };

  static defaultProps = {
    animate: autoPlayGif,
    size: 20,
    inline: false,
  };

  state = {
    hovering: false,
  };

  handleMouseEnter = () => {
    if (this.props.animate) return;
    this.setState({ hovering: true });
  };

  handleMouseLeave = () => {
    if (this.props.animate) return;
    this.setState({ hovering: false });
  };

  render () {
    const { account, size, animate, inline } = this.props;
    const { hovering } = this.state;

    const style = {
      ...this.props.style,
      width: `${size}px`,
      height: `${size}px`,
    };

    let src;

    if (hovering || animate) {
      src = account?.get('avatar');
    } else {
      src = account?.get('avatar_static');
    }

    return (
      <div className={classNames('account__avatar', { 'account__avatar-inline': inline })} onMouseEnter={this.handleMouseEnter} onMouseLeave={this.handleMouseLeave} style={style}>
        {src && <img src={src} alt={account?.get('acct')} />}
      </div>
    );
  }

}
