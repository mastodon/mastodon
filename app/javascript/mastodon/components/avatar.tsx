import classNames from 'classnames';

import type { Account } from 'mastodon/models/account';

import { useHovering } from '../../hooks/useHovering';
import { autoPlayGif } from '../initial_state';

interface Props {
  account: Account | undefined; // FIXME: remove `undefined` once we know for sure its always there
  size: number;
  style?: React.CSSProperties;
  inline?: boolean;
  animate?: boolean;
  counter?: number | string;
  counterBorderColor?: string;
}

export const Avatar: React.FC<Props> = ({
  account,
  animate = autoPlayGif,
  size = 20,
  inline = false,
  style: styleFromParent,
  counter,
  counterBorderColor,
}) => {
  const { hovering, handleMouseEnter, handleMouseLeave } = useHovering(animate);

  const style = {
    ...styleFromParent,
    width: `${size}px`,
    height: `${size}px`,
  };

  const src =
    hovering || animate
      ? account?.get('avatar')
      : account?.get('avatar_static');

  return (
    <div
      className={classNames('account__avatar', {
        'account__avatar-inline': inline,
      })}
      onMouseEnter={handleMouseEnter}
      onMouseLeave={handleMouseLeave}
      style={style}
    >
      {src && <img src={src} alt='' />}
      {counter && (
        <div
          className='account__avatar__counter'
          style={{ borderColor: counterBorderColor }}
        >
          {counter}
        </div>
      )}
    </div>
  );
};
