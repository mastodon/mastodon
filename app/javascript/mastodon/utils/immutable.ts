import { isImmutable } from 'immutable';

/**
 * Converts an immutable Map or Set to JS if it isn't already a plain JS object.
 * This is useful in code that needs to be able to deal with either type.
 */

export function immutableToJSIfNeeded<T>(object: T): T {
  return isImmutable(object) ? (object.toJS() as T) : object;
}
