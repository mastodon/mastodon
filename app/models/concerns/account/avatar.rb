# frozen_string_literal: true

module Account::Avatar
  extend ActiveSupport::Concern

  IMAGE_MIME_TYPES = ['image/jpeg', 'image/png', 'image/gif', 'image/webp'].freeze
  LIMIT = 2.megabytes

  AVATAR_DIMENSIONS = [400, 400].freeze
  AVATAR_GEOMETRY = [AVATAR_DIMENSIONS.first, AVATAR_DIMENSIONS.last].join('x')

  class_methods do
    def avatar_styles(file)
      styles = { original: { geometry: "#{AVATAR_GEOMETRY}#", file_geometry_parser: FastGeometryParser } }
      styles[:static] = { geometry: "#{AVATAR_GEOMETRY}#", format: 'png', convert_options: '-coalesce', file_geometry_parser: FastGeometryParser } if file.content_type == 'image/gif'
      styles
    end

    private :avatar_styles
  end

  included do
    # Avatar upload
    has_attached_file :avatar, styles: ->(f) { avatar_styles(f) }, convert_options: { all: '+profile "!icc,*" +set date:modify +set date:create +set date:timestamp' }, processors: [:lazy_thumbnail]
    validates_attachment_content_type :avatar, content_type: IMAGE_MIME_TYPES
    validates_attachment_size :avatar, less_than: LIMIT
    remotable_attachment :avatar, LIMIT, suppress_errors: false
  end

  def avatar_original_url
    avatar.url(:original)
  end

  def avatar_static_url
    avatar_content_type == 'image/gif' ? avatar.url(:static) : avatar_original_url
  end
end
