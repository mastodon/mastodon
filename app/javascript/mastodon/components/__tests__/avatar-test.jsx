import { fromJS } from 'immutable';

import renderer from 'react-test-renderer';

import { Avatar } from '../avatar';

describe('<Avatar />', () => {
  const account = fromJS({
    username: 'alice',
    acct: 'alice',
    display_name: 'Alice',
    avatar: '/animated/alice.gif',
    avatar_static: '/static/alice.jpg',
  });

  const size     = 100;

  describe('Autoplay', () => {
    it('renders a animated avatar', () => {
      const component = renderer.create(<Avatar account={account} animate size={size} />);
      const tree      = component.toJSON();

      expect(tree).toMatchSnapshot();
    });
  });

  describe('Still', () => {
    it('renders a still avatar', () => {
      const component = renderer.create(<Avatar account={account} size={size} />);
      const tree      = component.toJSON();

      expect(tree).toMatchSnapshot();
    });
  });

  // TODO add autoplay test if possible
});
