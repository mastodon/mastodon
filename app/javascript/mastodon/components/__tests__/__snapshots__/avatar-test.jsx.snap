// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Avatar /> Autoplay renders a animated avatar 1`] = `
<div
  className="account__avatar"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  style={
    {
      "height": "100px",
      "width": "100px",
    }
  }
>
  <img
    alt=""
    src="/animated/alice.gif"
  />
</div>
`;

exports[`<Avatar /> Still renders a still avatar 1`] = `
<div
  className="account__avatar"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  style={
    {
      "height": "100px",
      "width": "100px",
    }
  }
>
  <img
    alt=""
    src="/static/alice.jpg"
  />
</div>
`;
